lib/compress.js

```diff
@@ -652,7 +652,7 @@ merge(Compressor.prototype, {
             }
             if (parent instanceof AST_Assign && parent.operator == "=" && node === parent.right
                 || parent instanceof AST_Call && node !== parent.expression
-                || parent instanceof AST_Return && node === parent.value && node.scope !== d.scope
+                || parent instanceof AST_Exit && node === parent.value && node.scope !== d.scope
                 || parent instanceof AST_VarDef && node === parent.value) {
                 d.escaped = true;
```

Comment:
@alexlamsl For `harmony` we have to add escape support for `AST_Yield` and `AST_Await` as they are not derived from `AST_Exit`. 

And please don't try to change their base class to `AST_Exit` - I initially tried that and it failed miserably due to a myriad of implicit assumptions in the code base. The `yield` and `await` code works and is not as well tested as ES5 is with the fuzzer, so I'd prefer to minimize disturbing it.

Reply:
From my brief reading on MDN, `AST_Await` should be placed alongside `AST_Array`
https://github.com/mishoo/UglifyJS2/blob/74a2f53683e2b8788eadc10762583491e3dbc7ea/lib/compress.js#L659-L662

... whereas `AST_Yield` behaves like `AST_Exit` above as you suggested.

Reply:
@alexlamsl I created #2565 for harmony `yield`, but I was unable to craft a failing case for `async`/`await`. Care to give it a shot? Perhaps it's fine.

Reply:
> AST_Await should be placed alongside AST_Array

Are you referring to operator precedence? Yeah - `await` and `yield` are quite different as I discovered when implementing support for `async`/`await`.

> ... whereas AST_Yield behaves like AST_Exit above as you suggested.

We tried modelling `AST_Yield` in many different ways and the best way turned out just to derive it from `AST_Node` - no baggage or unwelcome assumptions in the then ES5-centric code base.

Reply:
I wasn't talking about `AST_Await` or `AST_Yield` in general AST terms, I was specifically referring to where to insert them within `mark_escaped()`.

You can give say `return await foo;` a go as it should execute identically as `return foo;` according to MDN.